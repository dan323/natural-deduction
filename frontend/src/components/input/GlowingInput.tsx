import React, { FC, useState, ChangeEventHandler } from 'react';
import './glowing.css'

type GlowingInputProps = {
    label: string;
    glowColor: string;
    shouldGlow: boolean;
    onColorChange: (color: string, line: number) => void;
    onInput: (index: number, input: number | string) => void;
    index: number;
};

function isNumeric(value: string) {
    return /^\d+$/.test(value);
}

const GlowingInput: FC<GlowingInputProps> = ({ label, glowColor, shouldGlow, onColorChange, onInput, index}) => {
    const [value, setValue] = useState<string>('');

    const inputId = `glowing-input-${index}`;
    const handleChange: ChangeEventHandler<HTMLInputElement> = (event) => {
        const inputValue = event.target.value;
        setValue(inputValue);
        if (isNumeric(inputValue)) {
            const numInput = parseInt(inputValue);
            onInput(index, numInput);
            // Trigger glow color change only for line number inputs
            if (shouldGlow && inputValue) {
                onColorChange(glowColor, numInput - 1);
            } else if (shouldGlow) {
                onColorChange(glowColor, -1);  // Remove glow when input is empty
            }
        } else if (inputValue === '') {
            onColorChange(glowColor, -1)
        } else {
            onInput(index, inputValue);
        }
    };

    return (
        <div className="input-item">
            <label htmlFor={inputId} className="input-label">{label}</label>
            <input
                id={inputId}
                type="text"
                className={`input-field ${shouldGlow && value ? 'glowing-input' : ''}`}
                style={{ boxShadow: value && shouldGlow ? `0 0 10px ${glowColor}, 0 0 40px ${glowColor}, 0 0 80px ${glowColor}` : 'none' }}
                onChange={handleChange}
                value={value}
            />
        </div>
    );
};

export default GlowingInput;