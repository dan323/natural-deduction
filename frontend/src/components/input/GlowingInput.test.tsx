import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import GlowingInput from './GlowingInput';

describe('Glowing input', () => {
    test('renders GlowingInput component', () => {
        const { getByLabelText } = render(
            <GlowingInput
                label="Enter line number:"
                glowColor="#ffcc00"
                shouldGlow={true}
                onColorChange={jest.fn()}
                onInput={jest.fn()}
                index={0}
            />
        );
        expect(getByLabelText(/Enter line number:/i)).toBeInTheDocument();
    });

    test('input glows when shouldGlow is true and value is present', () => {
        const { getByLabelText } = render(
            <GlowingInput
                label="Enter line number:"
                glowColor="#ffcc00"
                shouldGlow={true}
                onColorChange={jest.fn()}
                onInput={jest.fn()}
                index={0}
            />
        );

        const input = getByLabelText(/Enter line number:/i);
        act(() => {
            fireEvent.change(input, { target: { value: '1' } });
        });
        expect(input).toHaveStyle(`box-shadow: 0 0 10px #ffcc00, 0 0 40px #ffcc00, 0 0 80px #ffcc00`);
    });

    test('input does not glow when shouldGlow is false', () => {
        const { getByLabelText } = render(
            <GlowingInput
                label="Enter line number:"
                glowColor="#ffcc00"
                shouldGlow={false}
                onColorChange={jest.fn()}
                onInput={jest.fn()}
                index={0}
            />
        );

        const input = getByLabelText(/Enter line number:/i);
        act(() => {
            fireEvent.change(input, { target: { value: '1' } });
        });
        expect(input).not.toHaveStyle(`box-shadow: inset 0 0 10px #ffcc00`);
    });

    test('calls onInput with correct arguments when input changes', () => {
        const onInputMock = jest.fn();
        const { getByLabelText } = render(
            <GlowingInput
                label="Enter line number:"
                glowColor="#ffcc00"
                shouldGlow={true}
                onColorChange={jest.fn()}
                onInput={onInputMock}
                index={0}
            />
        );

        const input = getByLabelText(/Enter line number:/i);
        act(() => {
            fireEvent.change(input, { target: { value: '2' } });
        });
        expect(onInputMock).toHaveBeenCalledWith(0, 2);
    });

    test('calls onColorChange correctly when input value is valid', () => {
        const onColorChangeMock = jest.fn();
        const { getByLabelText } = render(
            <GlowingInput
                label="Enter line number:"
                glowColor="#ffcc00"
                shouldGlow={true}
                onColorChange={onColorChangeMock}
                onInput={jest.fn()}
                index={0}
            />
        );

        const input = getByLabelText(/Enter line number:/i);
        act(() => {
            fireEvent.change(input, { target: { value: '3' } });
        });
        expect(onColorChangeMock).toHaveBeenCalledWith("#ffcc00", 2);
    });

    test('does not apply glow or call onColorChange with non-numeric input', () => {
        const onColorChangeMock = jest.fn();
        const { getByLabelText } = render(
            <GlowingInput
                label="Enter line number:"
                glowColor="#ffcc00"
                shouldGlow={true}
                onColorChange={onColorChangeMock}
                onInput={jest.fn()}
                index={0}
            />
        );

        const input = getByLabelText(/Enter line number:/i);
        act(() => {
            fireEvent.change(input, { target: { value: 'abc' } });
        });
        expect(onColorChangeMock).not.toHaveBeenCalled();
        expect(input).not.toHaveStyle(`box-shadow: inset 0 0 10px #ffcc00`);
    });

    test('removes glow and calls onColorChange with -1 when input is cleared', () => {
        const onColorChangeMock = jest.fn();
        const { getByLabelText } = render(
            <GlowingInput
                label="Enter line number:"
                glowColor="#ffcc00"
                shouldGlow={true}
                onColorChange={onColorChangeMock}
                onInput={jest.fn()}
                index={0}
            />
        );

        const input = getByLabelText(/Enter line number:/i);
        act(() => {
            fireEvent.change(input, { target: { value: '5' } });
            fireEvent.change(input, { target: { value: '' } });
        })
        expect(onColorChangeMock).toHaveBeenCalledWith("#ffcc00", -1);
        expect(input).not.toHaveStyle(`box-shadow: inset 0 0 10px #ffcc00`);
    });

    test('matches snapshot when shouldGlow is true', () => {
        const { asFragment } = render(
            <GlowingInput
                label="Enter line number:"
                glowColor="#ffcc00"
                shouldGlow={true}
                onColorChange={jest.fn()}
                onInput={jest.fn()}
                index={0}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test('matches snapshot when shouldGlow is false', () => {
        const { asFragment } = render(
            <GlowingInput
                label="Enter line number:"
                glowColor="#ffcc00"
                shouldGlow={false}
                onColorChange={jest.fn()}
                onInput={jest.fn()}
                index={0}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});