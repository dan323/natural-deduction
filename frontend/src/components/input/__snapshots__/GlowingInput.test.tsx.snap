// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Glowing input matches snapshot when shouldGlow is false 1`] = `
<DocumentFragment>
  <div
    class="input-item"
  >
    <label
      class="input-label"
      for="glowing-input-0"
    >
      Enter line number:
    </label>
    <input
      class="input-field "
      id="glowing-input-0"
      style="box-shadow: none;"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`Glowing input matches snapshot when shouldGlow is true 1`] = `
<DocumentFragment>
  <div
    class="input-item"
  >
    <label
      class="input-label"
      for="glowing-input-0"
    >
      Enter line number:
    </label>
    <input
      class="input-field "
      id="glowing-input-0"
      style="box-shadow: none;"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
